src/main/scala/org/constellation/rewards/RewardsManager.scala

```diff
@@ -0,0 +1,219 @@
+
+class RewardsManager[F[_]: Concurrent](
+  eigenTrust: EigenTrust[F],
+  checkpointService: CheckpointService[F],
+  addressService: AddressService[F],
+  snapshotBroadcastService: SnapshotBroadcastService[F]
+) {
+  final val rewards: Ref[F, Map[String, Double]] = Ref.unsafe(Map.empty)
+
+  def updateBySnapshot(snapshot: Snapshot, snapshotHeight: Int): F[Unit] =
+    for {
+      observations <- observationsFromSnapshot(snapshot)
+      _ <- eigenTrust.retrain(observations)
+      trustMap <- eigenTrust.getTrustForAddressHashes
+
+      weightContributions = weightByTrust(trustMap) _ >>> weightByEpoch(snapshotHeight) >>> weightByStardust
+      contributions = calculateContributions(trustMap.keySet.toSeq)
+      distribution = weightContributions(contributions)
+
+      _ <- updateRewards(distribution)
+      _ <- updateAddressBalances(distribution)
+    } yield ()
+
+  def weightByStardust: Map[String, Double] => Map[String, Double] =
+    StardustCollective.weightByStardust
+
+  def weightByEpoch(snapshotHeight: Int)(contributions: Map[String, Double]): Map[String, Double] =
+    contributions.mapValues(_ * rewardDuringEpoch(snapshotHeight))
+
+  def weightByTrust(
+    trustEntropyMap: Map[String, Double]
+  )(contributions: Map[String, Double]): Map[String, Double] = {
+    val weightedEntropy = contributions.transform {
+      case (address, partitionSize) =>
+        partitionSize * (1 - trustEntropyMap(address)) // Normalize wrt total partition space
+    }
+    val totalEntropy = weightedEntropy.values.sum
+    weightedEntropy.mapValues(_ / totalEntropy) // Scale by entropy magnitude
+  }
+
+  /**
+    * Calculates non-weighted contributions so each address takes equal part which is calculated
+    * as 1/totalSpace
+    * @param addresses
+    * @return
+    */
+  def calculateContributions(addresses: Seq[String]): Map[String, Double] = {
+    val totalSpace = addresses.size
+    val contribution = 1.0 / totalSpace
+    addresses.map { address =>
+      address -> contribution
+    }.toMap
+  }
+
+  private def observationsFromSnapshot(snapshot: Snapshot): F[Seq[Observation]] =
+    snapshot.checkpointBlocks.toList
+      .traverse(checkpointService.fullData)
+      .map(_.flatten.flatMap(_.checkpointBlock.observations))
+
+  private def updateRewards(distribution: Map[String, Double]): F[Unit] =
+    rewards.modify(prevDistribution => (prevDistribution |+| distribution, ()))
+
+  /*
+    TODO: Not sure if it is the correct way of updating balances by rewards.
+    1. putUnsafe takes Account.hash but we pass there Id.address which could differ. Maybe we should do Address(id.address).hash.
+    2. I take the current balance and just copy the balance by adding rewards. Seems to be quite... dirty and unsafe?
+   */
+  private def updateAddressBalances(rewards: Map[String, Double]): F[Unit] =
+    rewards.transform {
+      case (address, reward) =>
+        addressService.lookup(address).flatMap { existingCacheData =>
```

Comment:
I made a comment on your other pr, but there needs to be a lock on memPool in addressService when updating balances.